src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -81,62 +81,55 @@ internal bool TryGetBuffer(in SequencePosition position, out ReadOnlyMemory<T> m
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in SequencePosition end)
+        private ReadOnlyMemory<T> GetFirstBuffer()
         {
-            SequenceType type;
-            int startIndex = 0;
-            int endIndex = 0;
-            object startObject = start.GetObject();
-            if (startObject != null)
-            {
-                GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out type, out startIndex, out endIndex);
-            }
-            else
-            {
-                type = SequenceType.Empty;
-            }
+            object startObject = _sequenceStart.GetObject();
 
-            int length = endIndex - startIndex;
-            object endObject = end.GetObject();
+            if (startObject == null)
+                return default;
 
-            if (type != SequenceType.MultiSegment && type != SequenceType.Empty && startObject != endObject)
-                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+            int startIndex = _sequenceStart.GetInteger();
+            int endIndex = _sequenceEnd.GetInteger();
 
-            ReadOnlyMemory<T> memory;
-            if (type == SequenceType.MultiSegment)
+            bool isMultiSegment = startObject != _sequenceEnd.GetObject();
+
+            if (startIndex >= 0)
             {
-                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
-                ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
-                memory = startSegment.Memory;
-                if (startObject != endObject)
+                if (endIndex >= 0)  // SequenceType.MultiSegment
```

Comment:
If we have to keep it for perf lets extract it to a method with clear name

Reply:
How big is performance difference if use GetSequenceType?

Reply:
> If we have to keep it for perf lets extract it to a method with clear name

What exactly would we extract into a method? Doing so would increase pressure on JIT to continue inlining.

I can revert back to using GetSequenceType, but I mainly structured the method this way to avoid any duplicate instructions (using GetSequenceType results in duplication the 'and' instructions to remove the bit mask). My assumption was that this method is super performance sensitive, so I was trying to get any savings possible.

If readability is important, I can make the trade off and go back to using GetSequenceType.

Reply:
There is significant perf hit if we use GetSequenceType:
![image](https://user-images.githubusercontent.com/6527137/38704744-2323b48a-3e5c-11e8-8d81-503b6253db82.png)

@pakrym, thoughts? Keep it as is, use GetSequenceType, extract to a method (can you expand on that)?

Reply:
Write a large comment above every `if` that checks the bit. Saying what exactly is it doing, why we don't use GetSeqType and such.

Reply:
Given we use GetSequenceType elsewhere, I am inclined to just use that instead of making this micro-optimization, and take the benefit of readability/maintainability.

Reply:
On the other hand perf improvement is pretty sweet

Reply:
I will add comments or use GetSequenceType in the next PR. I don't want to refresh the CI (it is unstable atm & takes a while and I want to get this change in soon).